README.md
# RESTful Natural Language Parse Service

A RESTful service that service NLP parsing requests.  This makes
the [NLP parsing project](https://github.com/plandes/clj-nlp-parse) available
as a [REST service](https://en.wikipedia.org/wiki/Representational_state_transfer).

<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-refresh-toc -->
## Table of Contents

- [Obtaining](#obtaining)
- [Documentation](#documentation)
- [Usage](#usage)
    - [Usage Help](#usage-help)
- [Docker](#docker)
- [Building](#building)
- [Changelog](#changelog)
- [License](#license)

<!-- markdown-toc end -->


## Obtaining

This project is designed to be used as a stand along application *and* a
library so you can add and use your own pipeline components.  For the library,
in your `project.clj` file, add:

[![Clojars Project](https://clojars.org/com.zensols.nlp/serv/latest-version.svg)](https://clojars.org/com.zensols.nlp/serv/)


## Documentation

Additional [documentation](https://plandes.github.io/clj-nlp-serv/codox/index.html).


## Usage

You can use this as a command line program to generate a prettyprint parse tree
of an utterance.  However, you have to let it know where the
Stanford CoreNLP libraries are configured (see the
[NLP parse project setup docs](https://github.com/plandes/clj-nlp-parse#setup)
for more information.

To create the command line utility, do the following:

- Install [Leiningen](http://leiningen.org) (this is just a script)
- Install [GNU make](https://www.gnu.org/software/make/)
- Install [Git](https://git-scm.com)
- Download the source:
```bash
$ git clone https://github.com/clj-nlp-parse
```
- Follow the directions in [build section](#building)
- Edit and uncomment the `makefile` to set the `ZMODEL` variable, which should
  be set to a directory having the stanford POS model(s) in `standford/pos`.
- Build the distribution binaries:
```bash
$ make dist
```
If everything goes well and you are lucky a new folder should show up on your
desktop with everything you need to run it.  To do that:
```bash
$ cd ~/Desktop/parse/bin
$ ./nlparse parse -u 'I am Paul Landes'
```

Now start the RESTful service, on say, port 9000 (defaults to 8080):
```bash
$ ./nlparse service -p 9000
```

Use a web client to test the service using the *pretty print* option:
```bash
$ wget -q -O - 'http://localhost:9000/parse?utterance=My+name+is+Paul+Landes&pretty=true'
```

The output:
```json
{"text":"My name is Paul Landes",
 "mentions":
 [{"entity-type":"PERSON",
   "token-range":[3, 5],
   "ner-tag":"PERSON",
   "sent-index":0,
   "char-range":[11, 22],
   "text":"Paul Landes"}],
 "tok-re-mentions":[],
 "coref":
 [{"id":1,
   "mention":
   [{"sent-index":1,
     "token-range":[4, 6],
     "head-index":5,
     "gender":"MALE",
     "animacy":"ANIMATE",
     "type":"PROPER",
     "number":"SINGULAR"}]},
```

**Note:** I will make the distribution binaries available on request.


### Usage Help

For convenience, here's the usage docs you get when invoking with no
parameters:
```sql
$ nlparse
repl	start a repl either on the command line or headless with -h
  -h, --headless            start an nREPL server
  -p, --port NUMBER  12345  the port bind for the repl server

 parse	parse an English utterance
  -l, --level LOG LEVEL  INFO  Log level to set in the Log4J2 system.
  -u, --utterance TEXT         The utterance to parse
  -p, --pretty                 Pretty print the JSON string

 service	start the agent parse website and service
  -p, --port PORT  8080  the port bind for web site/service

 version	Get the version of the application.
  -g, --gitref
```


## Docker

A [docker image is available](https://hub.docker.com/r/plandes/nlpservice/)
with an install of this service at [Docker Hub](http://hub.docker.com).  Here's
an example of how to use the image in a `docker-compose.yml`:
```yaml
nlpserv:
  container_name: nlps
  image: plandes/nlpservice
  ports:
    - "9100:9100"
  environment:
    COMPONENTS: tokenize,sentence,part-of-speech,morphology,stopword
    PORT: 9100
```


## Building

To build from source, do the folling:

- Install [Leiningen](http://leiningen.org) (this is just a script)
- Install [GNU make](https://www.gnu.org/software/make/)
- Install [Git](https://git-scm.com)
- Download the source: `git clone https://github.com/clj-nlp-serv && cd clj-nlp-serv`
- Download the make include files:
```bash
mkdir ../zenbuild && wget -O - https://api.github.com/repos/plandes/zenbuild/tarball | tar zxfv - -C ../zenbuild --strip-components 1
```
- Build the distribution binaries: `make dist`

Note that you can also build a single jar file with all the dependencies with: `make uber`


## Changelog

An extensive changelog is available [here](CHANGELOG.md).


## License

Copyright © 2016, 2017 Paul Landes

Apache License version 2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

[http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [Unreleased]

### Changed
- Move to new zenbuild.


## [0.0.9] - 2017-09-01
### Changed
- Tighter (unambiguous) and up to date dependencies.

## [0.0.8] - 2017-07-27
### Added
- Added annotations with version bump (see the [parse project] for more info):
  * Natural logic
  * Sentiment analysis
  
### Changed
- Fix internal server error.


## [0.0.7] - 2017-02-08
### Changed
- Version bump


[Unreleased]: https://github.com/plandes/clj-nlp-serv/compare/v0.0.9...HEAD
[0.0.9]: https://github.com/plandes/clj-nlp-serv/compare/v0.0.8...v0.0.9
[0.0.8]: https://github.com/plandes/clj-nlp-serv/compare/v0.0.7...v0.0.8
[0.0.7]: https://github.com/plandes/clj-nlp-serv/compare/v0.0.6...v0.0.7

[parse project]: https://github.com/plandes/clj-nlp-parse/blob/master/CHANGELOG.md


project.clj
(defproject com.zensols.nlp/serv "0.1.0-SNAPSHOT"
  :description "A RESTful service that service NLP parsing requests."
  :url "https://github.com/plandes/RESTful NLP parse service"
  :license {:name "Apache License version 2.0"
            :url "https://www.apache.org/licenses/LICENSE-2.0"
            :distribution :repo}
  :plugins [[lein-codox "0.10.1"]
            [org.clojars.cvillecsteele/lein-git-version "1.2.7"]]
  :codox {:metadata {:doc/format :markdown}
          :project {:name "RESTful NLP parse service"}
          :output-path "target/doc/codox"
          :source-uri "https://github.com/plandes/clj-nlp-serv/blob/v{version}/{filepath}#L{line}"}
  :git-version {:root-ns "zensols.nlserv"
                :path "src/clojure/zensols/nlserv"
                :version-cmd "git describe --match v*.* --abbrev=4 --dirty=-dirty"}
  :source-paths ["src/clojure"]
  :java-source-paths ["src/java"]
  :javac-options ["-Xlint:unchecked"]
  :jar-exclusions [#".gitignore"]
  :dependencies [[org.clojure/clojure "1.8.0"]

                 ;; logging
                 [org.apache.logging.log4j/log4j-core "2.7"]
                 [org.apache.logging.log4j/log4j-slf4j-impl "2.7"]
                 [org.apache.logging.log4j/log4j-1.2-api "2.7"]
                 [org.apache.logging.log4j/log4j-jcl "2.7"]

                 ;; web services
                 [ring/ring-core "1.5.0" :exclusions [joda-time]]
                 [ring/ring-jetty-adapter "1.5.0"]
                 [org.eclipse.jetty/jetty-server "9.2.10.v20150310"]
                 [compojure "1.5.1"]
                 [liberator "0.14.1"]

                 ;; json
                 [org.clojure/data.json "0.2.6"]

                 ;; command line
                 [com.zensols.nlp/parse "0.1.4"]]
  :pom-plugins [[org.codehaus.mojo/appassembler-maven-plugin "1.6"
                 {:configuration ([:programs
                                   [:program
                                    ([:mainClass "zensols.nlserv.core"]
                                     [:id "nlparse"])]]
                                  [:environmentSetupFileName "setupenv"])}]]
  :profiles {:1.9 {:dependencies [[org.clojure/clojure "1.9.0"]]}
             :uberjar {:aot [zensols.nlserv.core]}
             :appassem {:aot :all}
             :snapshot {:git-version {:version-cmd "echo -snapshot"}}
             :test
             {:jvm-opts
              ["-Dlog4j.configurationFile=test-resources/log4j2.xml"
               "-Xms4g" "-Xmx12g" "-XX:+UseConcMarkSweepGC"]}}
  :main zensols.nlserv.core)


src/clojure/zensols/nlserv/core.clj
(ns zensols.nlserv.core
  (:require [zensols.actioncli.parse :as cli]
            [zensols.actioncli.log4j2 :as lu])
  (:require [zensols.nlserv.version :as ver])
  (:gen-class :main true))

(def program-name "nlparse")

(defn- version-info-action []
  (println (format "%s (%s)" ver/version ver/gitref)))

(defn- create-action-context []
  (cli/multi-action-context
   '((:repl zensols.actioncli.repl repl-command)
     (:parse zensols.nlserv.handler parse-command)
     (:service zensols.nlserv.handler start-server-command)
     (:describe zensols.nlserv.handler describe-component-command))
   :version-option (cli/version-option version-info-action)))

(defn -main [& args]
  (lu/configure "nlp-serv-log4j2.xml")
  (cli/set-program-name program-name)
  (-> (create-action-context)
      (cli/process-arguments args)))


src/clojure/zensols/nlserv/service.clj
(ns zensols.nlserv.service
  (:require [clojure.data.json :as json]
            [clojure.tools.logging :as log])
  (:require [liberator.core :refer [defresource]])
  (:require [zensols.actioncli.dynamic :as dyn])
  (:require [zensols.nlparse.config :as conf :refer (with-context)]
            [zensols.nlparse.parse :as p]))

(def ^:private utterance-param "utterance")

(def ^:private pretty-param "pretty")

(defonce ^:private parse-context-inst (atom (conf/create-context)))

(defn set-parse-context [parse-context]
  (log/debugf "setting parse context: <%s>" (pr-str parse-context))
  (reset! parse-context-inst parse-context))

(defn- parse-context []
  @parse-context-inst)

(defn reset []
  (set-parse-context (conf/create-context)))

(dyn/register-purge-fn reset)

(defn- is-malformed?
  "Make sure the [[utterance-param]] key is given in the request parameters."
  [params]
  (not (contains? params utterance-param)))

(defn parse
  "Parse an human language utterance and return the JSON string."
  [utterance & {:keys [pretty?]}]
  (log/debugf "parsing (pretty?=%s) <%s>" pretty? utterance)
  (with-context (parse-context)
    (->> (p/parse utterance)
         ((if pretty?
            (fn [panon]
              (with-out-str (json/pprint panon)))
            json/write-str)))))

(defn- handle-parse-utterance
  "Service the parse utterance endpoint."
  [params]
  (let [pretty? (get params pretty-param)]
    (log/infof "servicing request with parameter: <%s>" params)
    (-> (get params utterance-param)
        (parse :pretty? pretty?))))

(defresource parse-utterance [params]
  :available-media-types ["text/json"]
  :allowed-methods [:post :get]
  :malformed? (fn [context]
                (let [params (get-in context [:request :params])]
                  (or (empty? params)
                      (is-malformed? params))))
  :handle-ok (fn [ctx]
               (handle-parse-utterance params))
  :handle-created (fn [ctx]
                    (handle-parse-utterance params))
  :post! (fn [ctx]
           (handle-parse-utterance params))
  :handle-not-found "Not found")


src/clojure/zensols/nlserv/handler.clj
(ns zensols.nlserv.handler
  (:require [clojure.tools.logging :as log]
            [clojure.string :as s]
            [clojure.java.io :as io])
  (:require [compojure.route :as route]
            [compojure.handler :as handler]
            [compojure.core :refer (defroutes GET POST ANY)]
            [ring.adapter.jetty :as jetty]
            [ring.middleware.params :refer [wrap-params]])
  (:require [zensols.actioncli.log4j2 :as lu]
            [zensols.actioncli.parse :refer (with-exception)]
            [zensols.nlparse.config :as conf])
  (:require [zensols.nlserv.service :as service]
            [zensols.nlserv.version :as ver]))

(defonce ^:private server-inst (atom nil))

(def ^:dynamic *dump-jvm-on-error* true)

(defroutes main-routes
  (GET "/ver/:type" [type]
       (if (= "git" type)
         ver/gitref
         ver/version))
  (ANY "/parse"
       {params :params}
       (service/parse-utterance params)))

(def ^:private app
  (-> main-routes
      (wrap-params)))

(defn- run-server
  ([]
   (run-server 8080))
  ([port]
   (log/infof "starting server of port: %d" port)
   (swap! server-inst
          #(or %
               (let [options {:port port :join? false}]
                 (jetty/run-jetty (var app) options))))))

(def ^:private components-option
  ["-c" "--components" (format "A comma separated list of pipeline components")
   :required "TEXT"])

(def describe-component-command
  "CLI command to print component documentation."
  {:description "component documentation"
   :options []
   :app (fn [opts & args]
          (conf/print-component-documentation))})

(def parse-command
  "CLI command to parse an utterance and return as a JSON string"
  {:description "parse an English utterance"
   :options
   [(lu/log-level-set-option)
    components-option
    ["-u" "--utterance" "The utterance to parse"
     :required "TEXT"
     :validate [#(> (count %) 0) "No utterance given"]]
    ["-p" "--pretty" "Pretty print the JSON string"]]
   :app (fn [{:keys [utterance components pretty] :as opts} & args]
          (with-exception
            (if (nil? utterance)
              (throw (ex-info "missing -u option" {:option "-u"})))
            (when components
              (->> (conf/create-context components)
                   service/set-parse-context))
            (println (service/parse utterance :pretty? pretty))))})

(def start-server-command
  {:description "start the agent parse website and service"
   :options
   [(lu/log-level-set-option)
    components-option
    ["-p" "--port PORT" "the port bind for web site/service"
     :default 9100
     :parse-fn #(Integer/parseInt %)
     :validate [#(< 0 % 0x10000) "Must be a number between 0 and 65536"]]]
   :app (fn [{:keys [components port]} & args]
          (when components
            (->> (conf/create-context components)
                 service/set-parse-context))
          (run-server port))})

;(run-server)
